### PR TITLE
Create file parser.

### PR DESCRIPTION
**Pull request by tpetersen0308**
This pull request creates a simple data file parsing command line program.

This code was originally written as a code challenge to be submitted in a .zip file, and was not committed at the time. Because this was written about a year ago, I am opening this PR to review my own code against my current standards for code design, with the intention of demonstrating the progress I have made as a developer since I wrote this program.

### CODE REVIEW

File: config/config.rb [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    comma: ",",
+    space: " "
+  },
+  gender_labels: {
+    male: ["M", "Male"],
+    female: ["F", "Female"]
+  },
+  input_types: [
+    {
+      delimiter: :pipe,
```
Comment (tpetersen0308) [inline config/config.rb:14]: I'm a little unclear on the intent of ```input_types``` here. Based on the README, it seems like this specifies the format of each file to be parsed, and as such the name seems a little vague.

Additionally, it looks like you're going for a declarative approach with this config setup, but is it getting a little too granular to have to specify each file individually? I don't think this will scale well.

File: config/config.rb [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+      column_order: [:last_name, :first_name, :gender, :birth_date, :favorite_color],
+      sort_rows_by: [
+        { key: :gender, direction: :asc },
+        { key: :last_name, direction: :asc }
+      ],
+      gender_labels: {
+        male: "Male",
+        female: "Female"
+      }
+    },
```
Comment (tpetersen0308) [inline config/config.rb:42]: In comparison to ```input_types```, this seems like a better use of config to me. It's relatively simple, and the user does not need to configure more than one type to use the program. I'd like to see ```input_types``` be more flexible in this way.

File: lib/app.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
+require_relative "./user_data_parser/parser.rb"
+
+def start(config, output_handler)
+  output_rulesets = config[:output_rulesets]
+  parser = UserDataParser::Parser.new(config)
+  output_rulesets.each do |rules|
+    parsed_data = parser.parse(rules)
+    output_handler.print(rules[:title])
+    parsed_data.each{ |row| output_handler.print(row) }
```
Comment (tpetersen0308) [inline lib/app.rb:9]: These nested loops make me nervous about scalability. I don't have any immediate suggestions for refactoring, but would like to discuss before assuming the technical debt.

File: lib/user_data_parser/output_handler/output_handler.rb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+module UserDataParser
+  class OutputHandler
+    def print(string)
+      puts string
```
Comment (tpetersen0308) [inline lib/user_data_parser/output_handler/output_handler.rb:4]: Is output handler just I/O? I'm glad you're thinking about dependency inversion, but this seems like it might be a premature abstraction. Moving forward, I think it's a good practice to wait until we see signs in the code that call for abstraction. It looks like this is only called in one file at the moment, so it's probably unnecessary. Let's leave it in though, but can you rename to something that calls out what it's abstracting a little more explicitly? Maybe even just ```IO```?

File: lib/user_data_parser/data_set/data_set.rb [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+    def initialize(rows = [])
+      @rows = rows
+    end
+
+    def add_rows(rows)
+      DataSet.new(@rows + rows)
+    end
+  end
+end
```
Comment (tpetersen0308) [inline lib/user_data_parser/data_set/data_set.rb:13]: See my comment below on ```OutputHandler```. I think a similar line of thinking may apply here. This looks like it might just be a fancy wrapper for an array.

File: lib/user_data_parser/data_set/data_set_parser.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    def initialize(row_parser)
+      @row_parser = row_parser
+    end
+
+    def parse(data)
+      data_set = DataSet.new
+      parsed_rows = data.map do |row|
+        row_parser.parse(row)
+      end
+      data_set = data_set.add_rows(parsed_rows)
```
Comment (tpetersen0308) [inline lib/user_data_parser/data_set/data_set_parser.rb:16]: Nitpick: this reads a little to me like we're altering a ```DataSet``` that already has rows. Why not just return ```DataSet.new(parsed_rows)```?

File: lib/user_data_parser/file/file_parser.rb [hunk 7]
```diff
@@ -0,0 +1,23 @@
+require_relative "../../../spec/config/test_config.rb"
+require_relative "../formatter.rb"
+require_relative "../row/row_parser.rb"
+require_relative "../data_set/data_set_parser.rb"
+
+module UserDataParser
```
Comment (tpetersen0308) [inline lib/user_data_parser/file/file_parser.rb:6]: Starting to get lost in the similarity of the names. I'm already seeing ```UserDataParser```, ```FileParser```, and ```DataSetParser```--when I read those, they all sound like they do the same thing. Can you think of a way to get more clarity out of the naming convention?

File: lib/user_data_parser/formatter.rb [hunk 8]
```diff
@@ -0,0 +1,48 @@
+require_relative "./row/row.rb"
+
+module UserDataParser
+  class Formatter
+    attr_reader :rules, :delimiters, :gender_labels
+
+    def initialize(rules, delimiters, gender_labels)
+      @rules = rules
+      @delimiters = delimiters
+      @gender_labels = gender_labels
```
Comment (tpetersen0308) [inline lib/user_data_parser/formatter.rb:10]: Why aren't ```delimiters``` and ```gender_labels``` simply a part of ```rules```?

File: lib/user_data_parser/formatter.rb [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+      delimiter = delimiters[rules[:delimiter]]
+      ordered_row.join(delimiter)
+    end
+
+    def format_row_attributes(row)
+      row_data = row.get_attributes
+      row_data[:birth_date] = format_date(row.birth_date)
+      row_data[:gender] = format_gender(row.gender)
+      row_data
+    end
```
Comment (tpetersen0308) [inline lib/user_data_parser/formatter.rb:25]: Nice, neat, readable methods. Good work!

File: lib/user_data_parser/parser.rb [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+      end
+
+      sort_key, sort_direction = [sorting_rules.first[:key], sorting_rules.first[:direction]]
+      sorted_rows = sort_rows(rows, sort_key, sort_direction)
+      chunked_rows = sorted_rows.chunk(&sort_key).to_a
+      sorted_row_chunks = chunked_rows.map{ |_value, row_chunk| row_chunk }
+
+      remaining_rules = sorting_rules[1..(sorting_rules.length - 1)]
+      sorted_row_chunks.map do |n|
+        chunk_sort(n, remaining_rules)
```
Comment (tpetersen0308) [inline lib/user_data_parser/parser.rb:42]: What is ```n```? Can we call this ```chunks``` or something more clear?

File: lib/user_data_parser/row/row.rb [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+module UserDataParser
+  class Row
+    def initialize(data)
+      set_columns(data)
+    end
+
+    def set_columns(data)
+      data.each do |k,v|
+        singleton_class.class_eval { attr_reader "#{k}" }
```
Comment (tpetersen0308) [inline lib/user_data_parser/row/row.rb:11]: Nitpick: consider encapsulating this in a private method so it's more clear what's going on here. What are we actually setting on the columns here?

File: lib/user_data_parser/row/row.rb [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+        attrs[var.to_s.delete("@").to_sym] = self.instance_variable_get(var)
+      end
+      attrs
+    end
+
+    private
+      def set_attribute(attr, value)
+        attr_name = "@#{attr}".to_sym
+        v = attr.to_s.include?("date") ? set_date(value) : value
+        self.instance_variable_set(attr_name, v)
```
Comment (tpetersen0308) [inline lib/user_data_parser/row/row.rb:28]: From [the docs](https://apidock.com/ruby/Object/instance_variable_set) on `instance_variable_set`:
> Sets the instance variable named by symbol to the given object, thereby frustrating the efforts of the class’s author to attempt to provide proper encapsulation. 

Do you think you can find a way to achieve similar behavior without metaprogramming?

File: lib/user_data_parser/row/row_parser.rb [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+    def initialize(delimiters, input_type)
+      @delimiters = delimiters
+      @input_type = input_type
+    end
+
+    def parse(data)
+      values = data.split(delimiters[input_type[:delimiter]])
+      parsed_row_data = {}
+      values.each_with_index do |v, i|
+        parsed_row_data[input_type[:column_order][i]] = v.strip
```
Comment (tpetersen0308) [inline lib/user_data_parser/row/row_parser.rb:16]: I think this calls out the vague intent of ```input_types``` that I brought up earlier. One thing that might mitigate it at least in this usage is to implement an explainer variable like
```column_name = input_type[:column_order][i]```

File: lib/user_data_parser/row/row_parser.rb [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+      return Row.new(parsed_row_data)
+    end
+
+    def parse_delimiter_type(data)
+      if data.include?(delimiters[:pipe])
+        return :pipe
+      elsif data.include?(delimiters[:comma])
+        return :comma
+      end
+      :space
```
Comment (tpetersen0308) [inline lib/user_data_parser/row/row_parser.rb:27]: Seems like we're pretty coupled to a specific set of delimiters here. Again, calls out the burden of effort placed on the user in configuring ```input_types```. I'd like to see this program simply reference ```delimiters``` and remain agnostic about what's inside. That could open up the possibility of users specifying insensible delimiters, like a vowel or something, but at that point it may be the user's prerogative.

File: spec/file/file_parser_spec.rb [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+      ]
+
+      actual = file_parser.parse("pipe.txt", data_set_parser)
+
+      expect(actual[0]).to have_attributes(expected[0])
+      expect(actual[1]).to have_attributes(expected[1])
+      expect(actual[2]).to have_attributes(expected[2])
+    end
+  end
+end
```
Comment (tpetersen0308) [inline spec/file/file_parser_spec.rb:46]: Another indicator of over-abstraction. This test is nearly identical to that of `data_set_parser`.

File: spec/formatter_spec.rb [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+      gender_labels = TEST_CONFIG[:gender_labels]
+      data = {
+        first_name: "Dale",
+        last_name: "Cooper",
+        gender: "M",
+        birth_date: "4-19-1954"
+      }
+      row = UserDataParser::Row.new(data)
+      formatter = UserDataParser::Formatter.new(rules, TEST_CONFIG[:delimiters], gender_labels)
+      expected = "4/19/1954 Male Cooper Dale"
```
Comment (tpetersen0308) [inline spec/formatter_spec.rb:18]: The test config reduces duplication in the tests, but it's also obscuring the connection between the configured rules and the expected result. Do you think that's worth the trade off here?

File: spec/integration_spec.rb [hunk 17]
```diff
@@ -0,0 +1,14 @@
...
+require_relative "./stubs/output_handler_stub.rb"
+require_relative "./stubs/output_stub.rb"
+
+RSpec.describe "App" do
+  it "parses and prints all data according to the config" do
+    output_handler = OutputHandlerStub.new
+    start(CONFIG, output_handler)
+    expected = OUTPUT_STUB
+
+    expect(output_handler.output).to eq(expected)
```
Comment (tpetersen0308) [inline spec/integration_spec.rb:12]: On a similar note to my comment above, it's really hard to understand what's actually under test here.

File: spec/row/row_parser_spec.rb [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+        last_name: "Cooper",
+        gender: "Male",
+        birth_date: set_date("4/19/1954")
+      }
+
+      actual = row_parser.parse(data)
+      expect(actual).to have_attributes(expected_attributes)
+    end
+
+    it "Parses space-delimited row of data according to the config" do
```
Comment (tpetersen0308) [inline spec/row/row_parser_spec.rb:41]: Wouldn't it be rad if these tests could just be "Parses rows based on provided delimiter and rules"?

File: spec/stubs/output_handler_stub.rb [hunk 19]
```diff
@@ -0,0 +1,11 @@
...
+  attr_accessor :output
+
+  def initialize
+    self.output = ""
+  end
+
+  def print(string)
+    self.output = self.output + "\n" + string
+  end
+end
```
Comment (tpetersen0308) [inline spec/stubs/output_handler_stub.rb:11]: I may take back what I said about premature abstraction for this module. I do prefer that the tests can rely on something like this instead of [directly mocking library methods](https://www.markhneedham.com/blog/2009/12/13/tdd-only-mock-types-you-own/?utm_source=feedburner&utm_medium=feed&utm_campaign=Feed%3A+MarkNeedham+%28Mark+Needham%29).
TL;DR:

> Programmers should not write mocks for fixed types, such as those defined by the runtime or external libraries. Instead they should write thin wrappers to implement the application abstractions in terms of the underlying infrastructure. Those wrappers will have been defined as part of a need-driven test.

**Review by tpetersen0308 — COMMENTED** [review]
I know I left you a lot of critical feedback, but I think overall this is a pretty well thought out design. Some of what I'm sharing entails specific changes, but most of it is suggestions for design considerations to help you grow as a developer, since it looks like you're already growing in the right direction. Here's my general feedback for this PR:

You're clearly thinking a lot about abstraction and naming, but there are parts of the code where the abstractions start to seem contrived. I think this is evident in the ambiguity of your parser class names. It seems like you might just be erring on the side of abstracting when you're unsure. Try letting the abstraction wait until its absence is just starting to cause issues.

That said, I think a lot of your abstractions are pretty good, particularly on the method level. Methods are readable with meaningful variable names, and I appreciate the use of clearly names symbols in place of literals.

There's no validation or error handling layer, which opens a lot of possibility for poor user experience. This is immediately evident in the fact that we're only testing the happy path. Let's start with validating the config, and providing meaningful output when a file can't be parsed.

I think the issue that requires the most immediate action is with `input_types`. Let's take care of that before it snowballs into a less manageable amount of tech debt.
